discovery_engine_core/web-api2/src/middleware/json_error.rs

```diff
@@ -0,0 +1,144 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+use std::{cmp::Ordering, fmt::Debug, future::Future};
+
+use actix_web::{
+    body::{BoxBody, MessageBody},
+    dev::{Service, ServiceRequest, ServiceResponse},
+    http::{header::CONTENT_TYPE, StatusCode},
+    HttpMessage,
+    HttpResponse,
+    ResponseError,
+};
+use derive_more::Display;
+use futures_util::TryFutureExt;
+use serde_json::{json, Value};
+
+use crate::error::json_error::JsonErrorResponseBuilder;
+
+use super::tracing::RequestId;
+
+pub(crate) fn wrap_non_json_errors<S, B>(
+    request: ServiceRequest,
+    service: &S,
+) -> impl Future<Output = Result<ServiceResponse<BoxBody>, actix_web::Error>> + 'static
+where
+    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = actix_web::Error>,
+    S::Future: 'static,
+    B: MessageBody + Debug + 'static,
+{
+    let request_id = request
+        .extensions()
+        .get::<RequestId>()
+        .copied()
+        .unwrap_or_else(RequestId::missing);
+
+    service
+        .call(request)
+        .map_ok(move |resp| wrap_service_response(resp, request_id))
+        // note that endpoints _directly_ turn any `Err(..)` into an `Ok(err_resp)`,
+        // so we will only see middleware errors here, never endpoint errors
+        .map_err(move |resp| WrappedMiddlewareError::wrap(resp, request_id))
+}
+
+fn wrap_service_response<B: MessageBody + Debug + 'static>(
+    response: ServiceResponse<B>,
+    request_id: RequestId,
+) -> ServiceResponse<BoxBody> {
+    if is_wrapable_error(response.response()) {
+        let (request, response) = response.into_parts();
+        let (response, body) = response.into_parts();
+        let details = extract_message_as_details(body);
+        let response =
+            JsonErrorResponseBuilder::render(response.status().as_str(), request_id, &details)
+                .apply_to(response);
+        ServiceResponse::new(request, response)
+    } else {
+        response.map_into_boxed_body()
+    }
+}
+
+fn is_wrapable_error<B>(response: &HttpResponse<B>) -> bool {
+    let status = response.status();
+    (status.is_client_error() || status.is_server_error())
+        && response
+            .headers()
+            .get(CONTENT_TYPE)
+            .map_or(true, |content_type| {
+                let mime_bytes = mime::TEXT_PLAIN.as_ref().as_bytes();
+                let bytes = content_type.as_bytes();
+                match bytes.len().cmp(&mime_bytes.len()) {
+                    Ordering::Less => false,
+                    Ordering::Equal => bytes == mime_bytes,
+                    Ordering::Greater => {
+                        bytes.starts_with(mime_bytes) && bytes[mime_bytes.len()] == b';'
+                    }
+                }
```

Comment:
I think it would be easier to understand with either an explanation comment or specifying explicit types

Reply:
yes it might be a bit of premature optimization, probably should just check equal `mime` or starts with `mime_with_trailing_semicolon`.